connection_generator.py

```diff
@@ -0,0 +1,197 @@
+            element_count: Total number of elements.
+            element_count_map: Dictionary that maps element count in a group to number of groups with that count.
+
+        Returns:
+            Dictionary that maps group id to a list of element ids.
+        """
+        # Create element ids
+        element_values = list(range(1, element_count + 1))
+
+        # Get number of elements for each group id from their count
+        elements_per_group = [i for i in element_count_map for _ in range(element_count_map[i])]
+
+        # Randomise element ids and group ids
+        random.shuffle(element_values)
+        random.shuffle(elements_per_group)
+
+        it = iter(element_values)
+
+        # Split element ids into chunks to get connections for each group
+        group_to_elements = {i + 1: set(islice(it, 0, elements_per_group[i])) for i in range(len(elements_per_group))}
+
+        return group_to_elements
+
+    @staticmethod
+    def get_many_to_many_connections(element_1_count, element_2_count, element_1_count_map, element_2_count_map):
+        """Generates random connections between elements of type 1 and type 2 that have many-to-many relationship.
+        Generation is based on element count maps. The output distribution is expected to be exact for most counts,
+        except for large element group outliers.
+
+        Args:
+            element_1_count: Total number of elements of type 1.
+            element_2_count: Total number of elements of type 2.
+            element_1_count_map: Dictionary that maps element 1 count in element 2 group to number of elements 2.
+            element_2_count_map: Dictionary that maps element 2 count in element 1 group to number of elements 1.
+
+        Returns:
+            Dictionary that maps group 1 id to a list of group 2 ids.
+        """
+        # Count zeros for each group
+        element_1_zeros = element_1_count_map[0] if 0 in element_1_count_map else 0
+        element_2_zeros = element_2_count_map[0] if 0 in element_2_count_map else 0
+
+        # Create element ids
+        element_1_values = list(range(1, element_1_count - element_1_zeros + 1))
+        element_2_values = list(range(1, element_2_count - element_2_zeros + 1))
+
+        # Get number of elements in each group and remove groups with 0 elements
+        elements_per_group_1 = [i for i in element_1_count_map for j in range(element_1_count_map[i]) if i != 0]
+        elements_per_group_2 = [i for i in element_2_count_map for j in range(element_2_count_map[i]) if i != 0]
+        element_1_group_counter = {i + 1: elements_per_group_1[i] for i in range(len(elements_per_group_1))}
+        element_2_group_counter = {i + 1: elements_per_group_2[i] for i in range(len(elements_per_group_2))}
+
+        # Create connection dictionary
+        element_1_conn_element_2 = {i: set() for i in element_1_values}
+
+        # Loop until run out of elements in any group
+        while element_1_values and element_2_values:
```

Comment:
How long does it take for this while loop to finish?

Reply:
Less than a second for a given config.